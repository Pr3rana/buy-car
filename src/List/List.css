.list{
    display: flex;
    border: 1px solid #ededed;
    margin: 12px;
}

.listContent{
    display: inline-block;
}

.product-icon{
    padding: 18px;
    height: 100px;
    width: 100px;
}
