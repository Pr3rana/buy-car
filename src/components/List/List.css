.list-item{
    display: flex;
    border: 1px solid #ededed;
    margin: 12px;
    align-items: center;
    padding: 12px;
    /* padding: 12px 0px 0px 12px; */
}

.list-header h3, .list-item-content h3{
    font-family: 'Roboto';
    font-weight: bold;
    font-size: 18px;
}

.list-header h3, .list-header p {
    margin: 0 12px 18px;
}

.list-item-content h3, .list-item-content p, .list-item-content a{
    margin: 0 12px 12px;
}

.list-item-content{
    display: inline-block;
    margin-left: 24px;
}

.item-key-details{
    font-size: 14px;
    font-weight: 600;
}

.product-icon{
    height: 80px;
    width: 80px;
}

.deatils-link{
    color: #ea7f28 !important;
    text-decoration: none;
    display: inline;
    font-size: 14px;
}

.deatils-link:hover, .deatils-link:active{
    color: #d37324;
    border-bottom: 1px solid #d37324;
}
