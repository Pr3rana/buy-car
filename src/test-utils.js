import React from 'react';
import { render } from '@testing-library/react'
import { BrowserRouter as Router, Route, Switch, Link } from 'react-router-dom'; 
import { ListPageContext } from "./helpers/appContext";

export const carData = {
    cars: [{"stockNumber":10000,"manufacturerName":"Tesla","modelName":"Model S","color":"silver","mileage":{"number":151884,"unit":"km"},"fuelType":"Diesel","pictureUrl":"https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"}],
    totalPageCount:100,
    totalCarsCount:1000
}
export const isFetched = false;
const setTotalCarsCount = jest.fn();
const setTotalPageCount = jest.fn();
const setPagenumber = jest.fn();
const setSelectedFilter = jest.fn();
const setSavedCars = jest.fn();
const setAvailableFilters = jest.fn();
const setFilterParams = jest.fn();

const pageContextProvider = {totalCarsCount:1,
    setTotalCarsCount,
    totalPageCount:2,
    setTotalPageCount,
    pagenumber:1,
    setPagenumber,
    selectedFilter:[{}],
    setSelectedFilter,
    savedCars:[],
    setSavedCars,
    availableFilters:{test:["test value"]},
    setAvailableFilters,
    filterParams: [{test:"test filter value"}],
    setFilterParams
}

const AllTheProviders = ({ children }) => {
    return (
     <Router>
          <ListPageContext.Provider value={pageContextProvider}>{children}</ListPageContext.Provider>
     </Router>
    )
}

const customRender = (ui, options) => {
    render(ui, { wrapper: AllTheProviders, ...options });
};

export * from "@testing-library/react";
export { customRender as render };