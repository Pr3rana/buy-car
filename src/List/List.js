import { useContext } from 'react';
import {Link} from 'react-router-dom';
import Pagination from '../Pagination/Pageination';
import { ListPageContext } from "../helpers/storeContext";
import './List.css';

function List({cars}){
    const { pagenumber } = useContext(ListPageContext);
    return ( 
        <div className="listContainer list-wrapper">
            <div>
                <h3>Available Cars</h3>
                <p>Showing {pagenumber*10} out of 100 results</p>
            </div>
            {cars.map((car,index)=>(
                <div className="list" key={index}>
                    <img alt="alt-text" className="product-icon" src={car.pictureUrl}/>
                    <div className="listContent">
                        <h3>{car.modelName}</h3>
                        <p className="product-key-details">
                            <span>{car.stockNumber}-</span>
                            <span>{car.mileage.number + car.mileage.unit}-</span>
                            <span>{car.fuelType}-</span>
                            <span>{car.color}</span>
                        </p>
                        <Link to = {`/details/${car.stockNumber}`}>
                            <p className="deatils-link">View details</p>
                        </Link>
                    </div>
                </div>
            ))}
             <Pagination/>
        </div>
     );
}
 
export default List;
