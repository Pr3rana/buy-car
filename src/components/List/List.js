import './List.css';
import { useContext, useEffect, useState } from 'react';
import {Link} from 'react-router-dom';
import Pagination from '../Pagination/Pageination';
import { ListPageContext } from "../../helpers/appContext";

function List({cars=[]}){
    const [availableCarCount, setAvailableCarCount] = useState(null);
    const [carCountLimit, setCarCountLimit] = useState(null);
    const { pagenumber, totalCarsCount } = useContext(ListPageContext);
    
    useEffect(()=>{
        setCarCountLimit(totalCarsCount > 100 ? 100 : totalCarsCount);
        setAvailableCarCount(carCountLimit <= 10 ? carCountLimit : (pagenumber*10) < carCountLimit ? (pagenumber*10) : carCountLimit);
    },[carCountLimit,pagenumber,totalCarsCount])

    return ( 
        cars.length>0 && <div data-testid="list-wrapper" className="list-container list-wrapper">
            <div className="list-header">
                <h3>Available Cars</h3>
                <p>Showing {availableCarCount} out of {carCountLimit} results</p>
            </div>
            {cars.map((car,index)=>(
                <div className="list-item" key={index}>
                    <img alt="alt-text" className="product-icon" src={car.pictureUrl}/>
                    <div className="list-item-content">
                        <h3>{car.modelName}</h3>
                        <p className="item-key-details">
                            <span>{car.stockNumber} - </span>
                            <span>{car.mileage.number +" " + car.mileage.unit} - </span>
                            <span>{car.fuelType} - </span>
                            <span>{car.color}</span>
                        </p>
                        <Link data-testid="deatils-link" className="deatils-link" to = {`/details/${car.stockNumber}`}>
                            <span>View details</span>
                        </Link>
                    </div>
                </div>
            ))}
            <Pagination/>
        </div>
     );
}
 
export default List;
