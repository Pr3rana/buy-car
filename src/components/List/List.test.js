import { render, screen, cleanup, carData } from '../../test-utils.js';
import List from './List';

afterEach(cleanup)
describe("List render",()=>{
    test('Initial render', ()=>{
        render(<List />)
        expect(screen.queryByTestId('list-wrapper')).toBeNull()
    })
})

describe("List props",()=>{
    test("testing list with props",()=>{
        render(<List cars={carData.cars} />)
        expect(screen.getByTestId('list-wrapper')).toMatchSnapshot(
            ` 
            <div className="list-container list-wrapper">
                <div class="list-header"><h3>Available Cars</h3><p>Showing 1 out of 1 results</p></div>
                <div class="list-item">
                <img alt="alt-text" class="product-icon" src="https://auto1-js-task-api--mufasa71.repl.co/images/car.svg">
                <div class="list-item-content"><h3>Model S</h3><p class="item-key-details"><span>10000 - </span>
                <span>151884 km - </span><span>Diesel - </span><span>silver</span></p><a class="deatils-link" href="/details/10000">
                <span>View details</span></a></div></div>
                <div data-testid="pagination-container" class="pagination-container"><span data-testid="first-page">First</span>
                <span data-testid="prev-page">Previous</span><span data-testid="curr-total-page">Page 1 of 2</span>
                <span data-testid="next-page">Next</span><span data-testid="last-page">Last</span></div>
            </div>
            `
        )
    })
})