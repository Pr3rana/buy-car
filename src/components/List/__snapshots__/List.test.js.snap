// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List props testing list with props:  
            <div className="list-container list-wrapper">
                <div class="list-header"><h3>Available Cars</h3><p>Showing 1 out of 1 results</p></div>
                <div class="list-item">
                <img alt="alt-text" class="product-icon" src="https://auto1-js-task-api--mufasa71.repl.co/images/car.svg">
                <div class="list-item-content"><h3>Model S</h3><p class="item-key-details"><span>10000 - </span>
                <span>151884 km - </span><span>Diesel - </span><span>silver</span></p><a class="deatils-link" href="/details/10000">
                <span>View details</span></a></div></div>
                <div data-testid="pagination-container" class="pagination-container"><span data-testid="first-page">First</span>
                <span data-testid="prev-page">Previous</span><span data-testid="curr-total-page">Page 1 of 2</span>
                <span data-testid="next-page">Next</span><span data-testid="last-page">Last</span></div>
            </div>
             1`] = `
<div
  class="list-container list-wrapper"
  data-testid="list-wrapper"
>
  <div
    class="list-header"
  >
    <h3>
      Available Cars
    </h3>
    <p>
      Showing 
      1
       out of 
      1
       results
    </p>
  </div>
  <div
    class="list-item"
  >
    <img
      alt="alt-text"
      class="product-icon"
      src="https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
    />
    <div
      class="list-item-content"
    >
      <h3>
        Model S
      </h3>
      <p
        class="item-key-details"
      >
        <span>
          10000
           - 
        </span>
        <span>
          151884 km
           - 
        </span>
        <span>
          Diesel
           - 
        </span>
        <span>
          silver
        </span>
      </p>
      <a
        class="deatils-link"
        data-testid="deatils-link"
        href="/details/10000"
      >
        <span>
          View details
        </span>
      </a>
    </div>
  </div>
  <div
    class="pagination-container"
    data-testid="pagination-container"
  >
    <span
      data-testid="first-page"
    >
      First
    </span>
    <span
      data-testid="prev-page"
    >
      Previous
    </span>
    <span
      data-testid="curr-total-page"
    >
      Page 
      1
       of 
      2
    </span>
    <span
      data-testid="next-page"
    >
      Next
    </span>
    <span
      data-testid="last-page"
    >
      Last
    </span>
  </div>
</div>
`;
